File: w10_d1_0_white_boarding/solutions/2_greatest_diff.rb
# Problem 2 - Greatest Difference

# Solution 1 - nested for loops (O(n^2) time)
def greatest_diff(array)
  max_diff = 0

  i = 0
  while (i < array.length)
    j = 0
    while (j < array.length)
      if array[i] - array[j] > max_diff
        max_diff = array[i] - array[j]
      end
      j += 1
    end
    i += 1
  end

  max_diff
end

# Solution 2 - single for loop (O(n) time)
def greatest_diff(array)
  min = Float::INFINITY
  max = -Float::INFINITY

  array.each do |num|
    if num < min
      min = num
    elsif num > max
      max = num
    end
  end

  max - min
end

File: w10_d1_0_white_boarding/solutions/2_greatest_diff.js
// Problem 2 - Greatest Difference

// Solution 1 - nested for loops (O(n^2) time)
function greatestDiff(array) {
  var maxDiff = 0;
  for (var i = 0; i < array.length; i++) {
    for (var j = 0; j < array.length; j++) {
      if (array[i] - array[j] > maxDiff) {
        maxDiff = array[i] - array[j];
      }
    }
  }
  return maxDiff;
}

// Solution 2 - single for loop (O(n) time)
function greatestDiff(array) {
  var min = Infinity;
  var max = - Infinity;
  for (var i = 0; i < array.length; i++) {
    if (array[i] < min) {
      min = array[i];
    }
    else if (array[i] > max) {
      max = array[i];
    }
  }
  return max - min;
}